win32ss/user/user32/misc/loadpic.c

```diff
@@ -0,0 +1,126 @@
+static ULONG_PTR s_dwGdiPlusToken   = 0;
+
+static FN_GdiplusStartup                s_fnStartup                 = NULL;
+static FN_GdiplusShutdown               s_fnShutdown                = NULL;
+static FN_GdipCreateBitmapFromFile      s_fnCreateBitmapFromFile    = NULL;
+static FN_GdipCreateHBITMAPFromBitmap   s_fnCreateHBITMAPFromBitmap = NULL;
+static FN_GdipDisposeImage              s_fnDisposeImage            = NULL;
+
+static BOOL IntLoadGdiPlus(VOID)
+{
+    HINSTANCE hGdiPlus;
+    GpStatus Error;
+    GdiplusStartupInput gdiplusStartupInput = { 1 };
+
+    if (s_hGdiPlus)
+        return TRUE;
+
+    hGdiPlus = LoadLibraryW(L"gdiplus.dll");
+    if (!hGdiPlus)
+        return FALSE;
+
+    s_fnStartup = (FN_GdiplusStartup)GetProcAddress(hGdiPlus, "GdiplusStartup");
+    s_fnShutdown = (FN_GdiplusShutdown)GetProcAddress(hGdiPlus, "GdiplusShutdown");
+    s_fnCreateBitmapFromFile =
+        (FN_GdipCreateBitmapFromFile)GetProcAddress(hGdiPlus, "GdipCreateBitmapFromFile");
+    s_fnCreateHBITMAPFromBitmap =
+        (FN_GdipCreateHBITMAPFromBitmap)GetProcAddress(hGdiPlus, "GdipCreateHBITMAPFromBitmap");
+    s_fnDisposeImage = (FN_GdipDisposeImage)GetProcAddress(hGdiPlus, "GdipDisposeImage");
+
+    if (!s_fnStartup || !s_fnShutdown ||
+        !s_fnCreateBitmapFromFile || !s_fnCreateHBITMAPFromBitmap || !s_fnDisposeImage)
+    {
+        FreeLibrary(hGdiPlus);
+        return FALSE;
+    }
+
+    Error = s_fnStartup(&s_dwGdiPlusToken, &gdiplusStartupInput, NULL);
+    if (Error)
+    {
+        ERR("%d\n", Error);
+        FreeLibrary(hGdiPlus);
+        return FALSE;
+    }
+
+    s_hGdiPlus = hGdiPlus;
+    return TRUE;
+}
+
+VOID IntFreeGdiPlus(VOID)
+{
+    if (s_hGdiPlus)
+    {
+        s_fnShutdown(s_dwGdiPlusToken);
+        s_dwGdiPlusToken = 0;
+
+        FreeLibrary(s_hGdiPlus);
+        s_hGdiPlus = NULL;
+    }
+}
+
+static HBITMAP IntLoadPicture(LPCWSTR pszFileName)
+{
+    HBITMAP hBitmap = NULL;
+    GpStatus Error;
+    GpBitmap *pBitmap;
+
+    TRACE("pszFileName: %ls\n", debugstr_w(pszFileName));
+
+    if (!pszFileName || !pszFileName[0] || !IntLoadGdiPlus())
+        return NULL;
+
+    Error = s_fnCreateBitmapFromFile(pszFileName, &pBitmap);
+    if (Error)
+        ERR("%d\n", Error);
+
+    Error = s_fnCreateHBITMAPFromBitmap(pBitmap, &hBitmap, 0xFFFFFFFF);
+    if (Error)
+        ERR("%d\n", Error);
+
+    Error = s_fnDisposeImage((GpImage*)pBitmap);
+    if (Error)
+        ERR("%d\n", Error);
+
+    return hBitmap;
+}
+
+NTSTATUS WINAPI
+User32CallLoadPictureFromKernel(PVOID Arguments, ULONG ArgumentLength)
+{
+    PLOADPICTURE_CALLBACK_ARGUMENTS Common = Arguments;
+    LOADPICTURE_CALLBACK_OUTPUT Result;
+    Result.hBitmap = IntLoadPicture(Common->szFileName);
+    return ZwCallbackReturn(&Result, sizeof(Result), STATUS_SUCCESS);
+}
+
+/* TODO: Support Vista PNG icons in the future,
+         using shlwapi!SHCreateMemStream, gdiplus!GdipCreateBitmapFromStream etc. */
```

Comment:
@Doug-Lyons You can insert the Vista icon related code here in the future. Just a note.